File: dotnet/cs/src/Modulation/RadiationPressure.cs
/*
 * File: RadiationPressure.cs
 * Project: Modulation
 * Created Date: 13/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 10/10/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
using Math = UnityEngine.Mathf;
#endif

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp.Modulation
{
    using Base = NativeMethods.Base;

    /// <summary>
    /// Modulation for modulating radiation pressure
    /// </summary>
    public sealed class RadiationPressure : Internal.Modulation
    {
        private readonly Internal.Modulation _m;

        public RadiationPressure(Internal.Modulation m)
        {
            _m = m;
        }

        public override ModulationPtr ModulationPtr()
        {
            return Base.AUTDModulationWithRadiationPressure(_m.ModulationPtr());
        }
    }

    public static class RadiationPressureModulationExtensions
    {
        public static RadiationPressure WithRadiationPressure(this Internal.Modulation s)
        {
            return new RadiationPressure(s);
        }
    }
}

File: python/pyautd3/modulation/radiation_pressure.py
"""
File: radiation_pressure.py
Project: modulation
Created Date: 10/10/2023
Author: Shun Suzuki
-----
Last Modified: 27/10/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


from typing import Generic, TypeVar

from pyautd3.internal.modulation import IModulation
from pyautd3.native_methods.autd3capi import NativeMethods as Base
from pyautd3.native_methods.autd3capi_def import ModulationPtr

M = TypeVar("M", bound=IModulation)


class RadiationPressure(IModulation, Generic[M]):
    """Modulation for modulating radiation pressure."""

    _m: M

    def __init__(self: "RadiationPressure", m: M) -> None:
        self._m = m

    def _modulation_ptr(self: "RadiationPressure") -> ModulationPtr:
        return Base().modulation_with_radiation_pressure(self._m._modulation_ptr())


def __with_radiation_pressure(self: M) -> RadiationPressure:
    """Apply modulation to radiation pressure instead of amplitude."""
    return RadiationPressure(self)


IModulation.with_radiation_pressure = __with_radiation_pressure  # type: ignore[method-assign]
